node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -334,12 +334,21 @@ public Node createNode(String openStackId, String hostname, Optional<String> par
 
     /** Adds a list of (newly created) nodes to the node repository as <i>provisioned</i> nodes */
     public List<Node> addNodes(List<Node> nodes) {
-        for (Node node : nodes) {
-            Optional<Node> existing = getNode(node.hostname());
-            if (existing.isPresent())
-                throw new IllegalArgumentException("Cannot add " + node.hostname() + ": A node with this name already exists");
-        }
         try (Mutex lock = lockAllocation()) {
+            for (int i = 0; i < nodes.size(); i++) {
+                var node = nodes.get(i);
+                var message = "Cannot add " + node.hostname() + ": A node with this name already exists";
+
+                // Check for existing node
+                if (getNode(node.hostname()).isPresent()) throw new IllegalArgumentException(message);
+
+                // Check for duplicates in given list
+                for (int j = 0; j < nodes.size(); j++) {
```

Comment:
You only need to check to `j < i` or start at `i + 1`, then you wont need the `if` below...